### ISSUE TITLE
[comparator] Disable foreign key constraint check/generation

### ISSUE THREAD

**Issue by quazardous** [doctrine/dbal#2525]
Hi,
(I'm coming from doctrine/migrations/issues#43)

I'm working with pretty big innodb tables and I need partitions more than foreign keys. Since they are mutually exclusive in MySQL I nee to disable foreign keys and put a more basic index instead.

So I would love to be able to explicitly disable foreign key check/generation with a cool annotation...

**Comment by duzymaju** [comment]
I agree with @quazardous - it would be great to have such functionality.

**Comment by JordyMoos** [comment]
Is there any work around to remove foreign keys on the partition table?

**Comment by mutale** [comment]
+1

**Comment by KarolyBujtor** [comment]
+1

**Comment by greg0ire** [comment]
If you need to voice your approval, please use the emoji reaction like the 12 other users that already did so. If you feel very strongly that this is needed, consider writing a PR, or drafting a plan to write a PR (look at the code, report here what would need to be done), but be constructive please.

**Comment by SenseException** [comment]
It is possible to use the [schema events](https://www.doctrine-project.org/projects/doctrine-dbal/en/2.9/reference/events.html#schema-events) of dbal to adapt the schema and therefore to remove the foreign keys from tables.

**Comment by fwolfsjaeger** [comment]
I had to disable creating foreign keys for certain tables as well. The simplest solution that I could find was overriding 2 methods of the Platform class.

I'm using Symfony, so I have created App\Doctrine\Platform\PostgreSQLPlatform:
```php
<?php

namespace App\Doctrine\Platform;

use Doctrine\DBAL\Platforms\PostgreSQLPlatform as DoctrinePostgreSQLPlatform;
use Doctrine\DBAL\Schema\TableDiff;

class PostgreSQLPlatform extends DoctrinePostgreSQLPlatform
{
    public function getCreateTablesSQL(array $tables): array
    {
        foreach ($tables as $table) {
            if (!str_starts_with($table->getName(), 'log_')) {
                continue;
            }

            $foreignKeys = $table->getForeignKeys();

            foreach ($foreignKeys as $foreignKey) {
                $table->removeForeignKey($foreignKey->getName());
            }
        }

        return parent::getCreateTablesSQL($tables);
    }

    public function getAlterTableSQL(TableDiff $diff): array
    {
        $tableName = $diff->getOldTable()?->getName();

        if (null !== $tableName && str_starts_with($tableName, 'log_')) {
            $diff->addedForeignKeys = [];
            $diff->changedForeignKeys = [];
            $diff->removedForeignKeys = [];
        }

        return parent::getAlterTableSQL($diff);
    }
}
```

Then you just have to configure the DBAL to use your platform:
```yaml
doctrine:
    dbal:
        connections:
            default:
                platform_service: App\Doctrine\Platform\PostgreSQLPlatform
```

**Comment by klobastov** [comment]
@fwolfsjaeger platform_service is deprecatad now :( 

` platform_service:     ~ # Deprecated (Since doctrine/doctrine-bundle 2.9: The "platform_service" configuration key is deprecated 
since doctrine-bundle 2.9. DBAL 4 will not support setting a custom platform via connection params anymore.)
`

**Comment by derrabus** [comment]
Wire a middleware that changes the platform.

**Comment by faizanakram99** [comment]
> It is possible to use the [schema events](https://www.doctrine-project.org/projects/doctrine-dbal/en/2.9/reference/events.html#schema-events) of dbal to adapt the schema and therefore to remove the foreign keys from tables.

Schema events don't exist in dbal v3 and onwards iirc. Whats the alternative? Custom platform?

That is doable but looks rather not scalable. 


Aaand custom platform doesn't work with non-default entity managers, at least not with doctrine migrations. I don't have a reproducer but I got bit by it pretty recently. Migrations bundle seems to use the platform that is shipped with doctrine  (it doesn't use the custom SchemaFactory) and doesn't respect doctrine bundle config.
